edx_django_utils/monitoring/internal/backends.py

```diff
@@ -0,0 +1,189 @@
+# and communication to anyone who might be specifically removing the
+# package for some reason.)
+#
+# Ticket for just doing an unconditional import:
+# https://github.com/openedx/edx-django-utils/issues/396
+try:
+    import newrelic.agent
+except ImportError:  # pragma: no cover
+    newrelic = None  # pylint: disable=invalid-name
+
+
+class TelemetryBackend(ABC):
+    """
+    Base class for telemetry sinks.
+    """
+    @abstractmethod
+    def set_attribute(self, key, value):
+        """
+        Set a key-value attribute on a span. This might be the current
+        span or it might the root span of the process, depending on
+        the backend.
+        """
+
+    @abstractmethod
+    def record_exception(self):
+        """
+        Record the exception that is currently being handled.
+        """
+
+
+class NewRelicBackend(TelemetryBackend):
+    """
+    Send telemetry to New Relic.
+
+    https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/guide-using-python-agent-api/
+    """
+    def __init__(self):
+        if newrelic is None:
+            raise Exception("Could not load New Relic monitoring backend; package not present.")
+
+    def set_attribute(self, key, value):
+        # Sets attribute on the transaction, rather than the current
+        # span, matching historical behavior. There is also an
+        # `add_custom_span_attribute` that would better match
+        # OpenTelemetry's behavior, which we could try exposing
+        # through a new, more specific TelemetryBackend method.
+        #
+        # TODO: Update to newer name `add_custom_attribute`
+        # https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/addcustomparameter-python-agent-api/
+        newrelic.agent.add_custom_parameter(key, value)
+
+    def record_exception(self):
+        # TODO: Replace with newrelic.agent.notice_error()
+        # https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/recordexception-python-agent-api/
+        newrelic.agent.record_exception()
+
+
+class OpenTelemetryBackend(TelemetryBackend):
+    """
+    Send telemetry via OpenTelemetry.
+
+    API reference: https://opentelemetry-python.readthedocs.io/en/latest/
+
+    Call ``OpenTelemetryBackend.initialize()`` from manage.py or gunicorn post-fork
+    hook.
+    """
+    initialized = False
+    init_lock = threading.Lock()
+
+    @classmethod
+    def initialize(cls):
+        """
+        Initialize and configure OpenTelemetry manually.
+
+        This is needed if not using opentelemetry-instrument server
+        startup wrapper (which is incompatible with gunicorn).
+        """
+        # Thread-safe one-time init to support gunicorn's forking model.
+        # (Otherwise we get warnings about re-initializing OTel.)
+        with cls.init_lock:
+            if cls.initialized:
+                return
+            else:
+                cls.initialized = True
```

Comment:
This still isn't quite working. Here's the output from gunicorn's post-fork hook on a sandbox, which ends up running this function twice:

```
2024-03-27 21:32:50,816 WARNING 3740339 [opentelemetry.trace] [user None] [ip None] __init__.py:521 - Overriding of current TracerProvider is not allowed
2024-03-27 21:32:50,846 WARNING 3740339 [opentelemetry.metrics._internal] [user None] [ip None] __init__.py:802 - Overriding of current MeterProvider is not allowed
2024-03-27 21:32:50,846 WARNING 3740339 [opentelemetry.instrumentation.instrumentor] [user None] [ip None] instrumentor.py:100 - Attempting to instrument while already instrumented
2024-03-27 21:32:50,905 WARNING 3740340 [opentelemetry.trace] [user None] [ip None] __init__.py:521 - Overriding of current TracerProvider is not allowed
2024-03-27 21:32:50,936 WARNING 3740340 [opentelemetry.metrics._internal] [user None] [ip None] __init__.py:802 - Overriding of current MeterProvider is not allowed
2024-03-27 21:32:50,936 WARNING 3740340 [opentelemetry.instrumentation.instrumentor] [user None] [ip None] instrumentor.py:100 - Attempting to instrument while already instrumented
```